sdks/go/pkg/beam/core/graph/mtime/time_test.go

```diff
@@ -0,0 +1,215 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mtime
+
+import (
+	"math"
+	"testing"
+	"time"
+)
+
+func TestAdd(t *testing.T) {
+	tests := []struct {
+		name     string
+		baseTime Time
+		addition time.Duration
+		expOut   Time
+	}{
+		{
+			"insignificant addition small",
+			Time(1000),
+			1 * time.Nanosecond,
+			Time(1000),
+		},
+		{
+			"insignificant addition large",
+			Time(1000),
+			999999 * time.Nanosecond,
+			Time(1000),
+		},
+		{
+			"significant addition small",
+			Time(1000),
+			1 * time.Millisecond,
+			Time(1001),
+		},
+		{
+			"significant addition large",
+			Time(1000),
+			10 * time.Second,
+			Time(11000),
+		},
+		{
+			"add past max timestamp",
+			MaxTimestamp,
+			1 * time.Minute,
+			MaxTimestamp,
+		},
+		{
+			"add across max boundary",
+			Time(int64(MaxTimestamp) - 10000),
+			10 * time.Minute,
+			MaxTimestamp,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if got, want := test.baseTime.Add(test.addition), test.expOut; got != want {
+				t.Errorf("got time %v, want %v", got, want)
```

Comment:
This is the kind of test that warrants having clarified inputs.

Same below for the the other tests.

```suggestion
				t.Errorf("(%v).Add(%v), got time %v, want %v", t.baseTime, test.addition, got, want)
```